File: android2/app/src/main/java/com/xbot/xbot/logging/PersonaLogServer.java
package com.xbot.xbot.logging;

import android.util.Log;

import androidx.annotation.Nullable;

import com.google.gson.Gson;

import java.io.IOException;
import java.net.Inet4Address;
import java.net.InetAddress;
import java.net.NetworkInterface;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Enumeration;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.regex.Pattern;

import fi.iki.elonen.NanoHTTPD;

/**
 * NanoHTTPD server exposing persona logs on the LAN (default port 8765).
 */
public class PersonaLogServer extends NanoHTTPD {
    private static final String TAG = "PersonaLogServer";
    private static final int DEFAULT_PORT = 8765;
    private static final Pattern DATE_PATTERN = Pattern.compile("^\\d{4}-\\d{2}-\\d{2}$");
    private static final Gson GSON = new Gson();

    private final PersonaLogger logger;
    @Nullable private String url;

    public PersonaLogServer(PersonaLogger logger) {
        super(DEFAULT_PORT);
        this.logger = logger;
    }

    public PersonaLogServer(PersonaLogger logger, int port) {
        super(port);
        this.logger = logger;
    }

    @Nullable
    public String getUrl() {
        return url;
    }

    public String startServer() throws IOException {
        if (!isAlive()) {
            start(NanoHTTPD.SOCKET_READ_TIMEOUT, false);
        }
        String ip = detectLanIpv4();
        url = "http://" + (ip != null ? ip : "localhost") + ":" + getListeningPort();
        Log.i(TAG, "started at " + url);
        return url;
    }

    public void stopServer() {
        stop();
        url = null;
        Log.i(TAG, "stopped");
    }

    @Override
    public Response serve(IHTTPSession session) {
        String path = session.getUri();
        Method method = session.getMethod();

        try {
            if ("/".equals(path) || "/index.html".equals(path)) {
                return withCors(newFixedLengthResponse(
                        Response.Status.OK,
                        "text/html; charset=utf-8",
                        dashboardHtml()));
            }
            if ("/healthz".equals(path)) {
                return withCors(jsonResponse("{\"ok\":true}"));
            }
            if ("/api/dates".equals(path)) {
                List<String> dates = awaitDates();
                Map<String, Object> body = new HashMap<>();
                body.put("dates", dates);
                body.put("dir", logger.getDirectoryPath());
                return withCors(jsonResponse(GSON.toJson(body)));
            }
            if ("/api/logs".equals(path)) {
                String date = session.getParms().get("date");
                if (date == null || !DATE_PATTERN.matcher(date).matches()) {
                    return withCors(jsonResponse(
                            "{\"error\":\"missing or invalid date\"}",
                            Response.Status.BAD_REQUEST));
                }
                if (Method.DELETE.equals(method)) {
                    CountDownLatch latch = new CountDownLatch(1);
                    logger.deleteDate(date, latch::countDown);
                    latch.await(5, TimeUnit.SECONDS);
                    return withCors(jsonResponse("{\"ok\":true,\"deleted\":\"" + date + "\"}"));
                }
                List<PersonaLogger.PersonaLogEntry> entries = awaitLogs(date);
                List<Map<String, Object>> logs = new ArrayList<>();
                for (PersonaLogger.PersonaLogEntry e : entries) {
                    logs.add(e.toJson());
                }
                Map<String, Object> body = new HashMap<>();
                body.put("date", date);
                body.put("count", logs.size());
                body.put("logs", logs);
                return withCors(jsonResponse(GSON.toJson(body)));
            }
            return withCors(newFixedLengthResponse(
                    Response.Status.NOT_FOUND,
                    MIME_PLAINTEXT,
                    "404 Not Found"));
        } catch (Exception e) {
            Log.e(TAG, "handle error", e);
            return withCors(newFixedLengthResponse(
                    Response.Status.INTERNAL_ERROR,
                    MIME_PLAINTEXT,
                    "500 Internal Server Error"));
        }
    }

    private List<String> awaitDates() throws InterruptedException {
        final List<String>[] holder = new List[1];
        CountDownLatch latch = new CountDownLatch(1);
        logger.availableDates(dates -> {
            holder[0] = dates;
            latch.countDown();
        });
        latch.await(5, TimeUnit.SECONDS);
        return holder[0] != null ? holder[0] : Collections.emptyList();
    }

    private List<PersonaLogger.PersonaLogEntry> awaitLogs(String date) throws InterruptedException {
        final List<PersonaLogger.PersonaLogEntry>[] holder = new List[1];
        CountDownLatch latch = new CountDownLatch(1);
        logger.readDate(date, entries -> {
            holder[0] = entries;
            latch.countDown();
        });
        latch.await(10, TimeUnit.SECONDS);
        return holder[0] != null ? holder[0] : Collections.emptyList();
    }

    private static Response jsonResponse(String json) {
        return jsonResponse(json, Response.Status.OK);
    }

    private static Response jsonResponse(String json, Response.Status status) {
        return newFixedLengthResponse(status, "application/json; charset=utf-8", json);
    }

    private static Response withCors(Response response) {
        response.addHeader("Access-Control-Allow-Origin", "*");
        return response;
    }

    @Nullable
    private static String detectLanIpv4() {
        try {
            Enumeration<NetworkInterface> interfaces = NetworkInterface.getNetworkInterfaces();
            while (interfaces != null && interfaces.hasMoreElements()) {
                NetworkInterface ni = interfaces.nextElement();
                if (!ni.isUp() || ni.isLoopback()) {
                    continue;
                }
                Enumeration<InetAddress> addrs = ni.getInetAddresses();
                while (addrs.hasMoreElements()) {
                    InetAddress addr = addrs.nextElement();
                    if (addr instanceof Inet4Address && !addr.isLoopbackAddress()) {
                        return addr.getHostAddress();
                    }
                }
            }
        } catch (Exception e) {
            Log.w(TAG, "lan ip detect failed", e);
        }
        return null;
    }

    private static String dashboardHtml() {
        return "<!DOCTYPE html><html><head><meta charset=\"utf-8\">"
                + "<title>XBot Persona Logs</title>"
                + "<style>body{font-family:sans-serif;margin:24px}"
                + "table{border-collapse:collapse;width:100%}"
                + "td,th{border:1px solid #ccc;padding:6px;font-size:13px}</style></head>"
                + "<body><h1>Persona Logs</h1>"
                + "<label>Date <select id=\"dates\"></select></label> "
                + "<button onclick=\"load()\">Load</button>"
                + "<table><thead><tr><th>Time</th><th>Type</th><th>Person</th>"
                + "<th>User</th><th>Reply</th></tr></thead><tbody id=\"rows\"></tbody></table>"
                + "<script>"
                + "async function init(){const r=await fetch('/api/dates');const j=await r.json();"
                + "const s=document.getElementById('dates');(j.dates||[]).forEach(d=>{"
                + "const o=document.createElement('option');o.value=d;o.textContent=d;s.appendChild(o)});"
                + "if(s.options.length)load();}"
                + "async function load(){const d=document.getElementById('dates').value;"
                + "const r=await fetch('/api/logs?date='+encodeURIComponent(d));const j=await r.json();"
                + "const tb=document.getElementById('rows');tb.innerHTML='';"
                + "(j.logs||[]).forEach(e=>{const tr=document.createElement('tr');"
                + "tr.innerHTML='<td>'+(e.ts||'')+'</td><td>'+(e.type||'')+'</td><td>'"
                + "+(e.person||'')+'</td><td>'+(e.userText||'')+'</td><td>'+(e.replyText||'')+'</td>';"
                + "tb.appendChild(tr)});}"
                + "init();</script></body></html>";
    }
}

File: xbot-android/app/src/main/java/com/xbot/android/log/PersonaLogServer.kt
package com.xbot.android.log

import android.content.Context
import android.util.Log
import fi.iki.elonen.NanoHTTPD
import java.util.Locale

/**
 * 人物日志 HTTP 浏览服务。对应 Flutter persona_log_server.dart。
 *
 * 局域网内电脑端浏览器访问，查看按天的人物行为日志和网络交互日志。
 * - GET /              日志首页（列出日期 + 链接）
 * - GET /day?d=yyyy-MM-dd  某天人物日志
 * - GET /net           网络交互日志列表
 * - GET /net?d=yyyy-MM-dd  某天网络日志原文
 *
 * 端口自动选择（:0 → 系统分配空闲端口）。
 */
class PersonaLogServer(
    private val context: Context,
    private val personaLogger: PersonaLogger,
    private val networkLogger: NetworkLogger,
) : NanoHTTPD(0) {

    companion object { private const val TAG = "PersonaLogServer" }

    @Volatile private var serverUrl: String? = null

    /** 启动并返回可访问 URL。 */
    fun startAndGetUrl(): String {
        serverUrl?.let { return it }
        start()
        val url = "http://localhost:$listeningPort"
        serverUrl = url
        Log.i(TAG, "日志服务启动: $url")
        return url
    }

    fun stopServer() {
        stop()
        serverUrl = null
    }

    override fun serve(session: IHTTPSession): Response {
        val path = session.uri
        val params = session.parameters
        return try {
            when {
                path == "/" -> newFixedLengthResponse(buildIndexPage())
                path == "/day" -> {
                    val day = params["d"]?.firstOrNull() ?: personaLogger.listDays().firstOrNull() ?: ""
                    val entries = if (day.isNotEmpty()) personaLogger.readDay(day) else emptyList()
                    newFixedLengthResponse(buildDayPage(day, entries))
                }
                path == "/net" -> {
                    val day = params["d"]?.firstOrNull() ?: networkLogger.listDays().firstOrNull() ?: ""
                    val raw = if (day.isNotEmpty()) networkLogger.readDay(day) else "（无记录）"
                    newFixedLengthResponse(buildNetPage(day, raw))
                }
                else -> newFixedLengthResponse(Response.Status.NOT_FOUND, "text/plain", "404")
            }
        } catch (e: Exception) {
            newFixedLengthResponse("Error: ${e.message}")
        }
    }

    private fun buildIndexPage(): String {
        val days = personaLogger.listDays()
        val links = days.joinToString("") { d -> """<li><a href="/day?d=$d">$d</a> | <a href="/net?d=$d">网络日志</a></li>""" }
        return """<!DOCTYPE html><html><head><meta charset="utf-8"><title>XBot 日志</title>
            <style>body{font-family:sans-serif;background:#1a1a2e;color:#e0e0e0;padding:20px}a{color:#80c0ff}li{margin:8px 0}</style>
            </head><body><h1>XBot 人物日志</h1><ul>$links</ul></body></html>"""
    }

    private fun buildDayPage(day: String, entries: List<PersonaLogEntry>): String {
        val rows = entries.joinToString("") { e ->
            val time = java.text.SimpleDateFormat("HH:mm:ss", Locale.getDefault()).format(java.util.Date(e.timestamp))
            """<tr><td>$time</td><td>${e.type}</td><td>${e.person ?: ""}</td><td>${e.expression ?: ""}</td><td>${e.gesture ?: ""}</td><td>${e.objects?.joinToString(",") ?: ""}</td><td>${e.heldObject ?: ""}</td><td>${e.behaviorState ?: ""}</td><td>${e.userText ?: ""}</td><td>${e.replyText ?: ""}</td></tr>"""
        }
        return """<!DOCTYPE html><html><head><meta charset="utf-8"><title>$day 日志</title>
            <style>body{font-family:monospace;background:#1a1a2e;color:#e0e0e0;padding:20px}table{border-collapse:collapse;width:100%}td,th{border:1px solid #333;padding:4px 8px;font-size:13px}th{background:#2a2a4e}</style>
            </head><body><h1>$day 人物日志</h1><p><a href="/">← 返回</a></p>
            <table><tr><th>时间</th><th>类型</th><th>人物</th><th>表情</th><th>手势</th><th>物体</th><th>手持</th><th>状态</th><th>用户说</th><th>机器人回复</th></tr>
            $rows</table></body></html>"""
    }

    private fun buildNetPage(day: String, raw: String): String {
        return """<!DOCTYPE html><html><head><meta charset="utf-8"><title>$day 网络日志</title>
            <style>body{font-family:monospace;background:#1a1a2e;color:#e0e0e0;padding:20px}pre{white-space:pre-wrap;word-break:break-all;font-size:12px}</style>
            </head><body><h1>$day 网络交互日志</h1><p><a href="/">← 返回</a></p>
            <pre>${raw.replace("<", "&lt;").replace(">", "&gt;")}</pre></body></html>"""
    }
}

private const val TAG = "PersonaLogServer"
